pricegraph/src/graph/path.rs

```diff
@@ -2,33 +2,54 @@
 
 use petgraph::graph::NodeIndex;
 
-/// Finds a cycle and returns a vector representing a path along the cycle,
-/// ending that is the predecessor of the starting node.
+/// Finds a cycle starting at the provided `search` node and returns a vector
+/// representing a path along the cycle.
+///
+/// Optionally, an `origin` node can be provided so that the first element of
+/// the cycle vector is `origin` if and only if `origin` is part of the cycle.
 ///
 /// Returns `None` if no such cycle can be found.
-pub fn find_cycle(predecessor: &[Option<NodeIndex>], start: NodeIndex) -> Option<Vec<NodeIndex>> {
+pub fn find_cycle(
+    predecessor: &[Option<NodeIndex>],
+    search: NodeIndex,
+    origin: Option<NodeIndex>,
+) -> Option<Vec<NodeIndex>> {
     // NOTE: First find a node that is actually on the cycle, this is done
     // because a negative cycle can be detected on any node connected to the
     // cycle and not just nodes on the cycle itself.
-    let mut visited = vec![false; predecessor.len()];
-    let mut current = start;
-    visited[current.index()] = true;
+    let mut visited = vec![0; predecessor.len()];
+    let mut cursor = search;
+    let mut step = 1;
+    visited[cursor.index()] = step;
     loop {
-        current = predecessor[current.index()]?;
-        if visited[current.index()] {
+        cursor = predecessor[cursor.index()]?;
+        if visited[cursor.index()] > 0 {
             break;
         }
-        visited[current.index()] = true;
+        step += 1;
+        visited[cursor.index()] = step;
     }
 
-    // NOTE: `current` is now guaranteed to be on the cycle, so just walk
-    // backwards until we reach `current` again.
-    let start = current;
-    let mut path = Vec::with_capacity(predecessor.len());
+    // NOTE: Allocate the cycle vector with enough capacity for the negative
+    // cycle plus one. This extra capacity is used by the orderbook graph to
+    // create a circular path equivalent to the negative cycle.
+    let len = step + 1 - visited[cursor.index()];
+    let mut path = Vec::with_capacity(len + 1);
```

Comment:
nit: Isn't it a bit hacky that we overallocate the vector for use in a specific case. Should this method really now how it's going to be used?

Reply:
Yeah, I can refactor this bit.

Reply:
I modified `find_path` to now return an actual path along the cycle, i.e. the first and last items are guaranteed to be the same.

I feel like this makes things less hacky since we are no longer over-allocating the vector, but instead just changed the semantics of the method. Note that `find_cycle` was only ever used this way, so it got rid of some `cycle.push(cycle[0])` which was already a bit ugly in my opinion.